alica_engine/include/engine/blackboard/Blackboard.h

```diff
@@ -23,25 +54,143 @@ class BlackboardImpl
+    std::unordered_map<std::string, BlackboardValue> vals;
+    YAML::Node node;
+
+private:
+    friend BlackboardUtil;
+    YAML::Node getBlackboardValueNode(std::string key) const
+    {
+        for (const auto& entry : node) {
+            if (entry[Strings::key].as<std::string>() == key) {
+                return entry;
+            }
+        }
+        return YAML::Node();
+    }
+
+    std::string getBlackboardValueType(const std::string& key) const
+    {
+        YAML::Node entry = getBlackboardValueNode(key);
+        return entry[Strings::stateType].as<std::string>();
+    }
+
+    template <class T>
+    std::string getNameFromType() const;
+};
+
+struct Converter
+{
+    static constexpr const char* typeNames[] = {"bool", "int64", "double", "std::string", "std::any"};
+
+    // YAML doesnt support std::any, needs a separate list of types for iteration
+    static constexpr const char* typeNamesYAML[] = {"bool", "int64", "double", "std::string"};
+    using TypesYAML = std::variant<bool, int64_t, double, std::string>;
+
+    // convert from type to name as string
+    template <class T>
+    static constexpr const char* typeName()
+    {
+        return typeNames[BlackboardValue{T{}}.index()];
+    }
+
+    // set the value given the type name as a string
+    static void setDefaultValue(const std::string& key, const std::string& typeName, const YAML::Node& defaultValue, BlackboardImpl& bb);
+
+    // used to set the value without a yaml node
+    static void setValue(const std::string& key, const BlackboardValue& value, const std::string& typeName, BlackboardImpl& bb);
+
+    template <class T>
+    struct YamlAs
+    {
+        static T as(const YAML::Node& value) { return value.as<T>(); }
+    };
+
+    template <class T, std::size_t INDEX, template <class> class Parser = YamlAs>
+    struct makeVariantIfEqual
+    {
+        static BlackboardValue make(std::size_t index, const YAML::Node& value)
+        {
+            if (index == INDEX) {
+                return Parser<T>::as(value);
+            }
+            return index == INDEX ? Parser<T>::as(value) : std::any{};
+        }
+    };
+
+    template <class... Ts, std::size_t... Is>
+    static BlackboardValue makeVariantFromIndex(std::variant<Ts...>, std::index_sequence<Is...>, std::size_t index, const YAML::Node& value)
+    {
+        BlackboardValue typeValues[] = {makeVariantIfEqual<Ts, Is>::make(index, value)...};
+        return typeValues[index];
+    };
+
+    template <class T, std::size_t INDEX>
+    struct setVariantIfEqual
+    {
+        static bool set(std::size_t index, const std::string& key, const BlackboardValue& value, BlackboardImpl& bb)
+        {
+            if (index != INDEX) {
+                return false;
+            }
+            const T& val = std::get<T>(value);
+            bb.set<T>(key, val);
+            return true;
+        }
+    };
+
+    template <class... Ts, std::size_t... Is>
+    static void setBlackboardValue(
```

Comment:
is this really required? Can't we just store the variant returned by `makeVariantFromIndex()` in the `vals` `unordered_map`?